# IBM Applied Data Science Capstone Project Course / Week 3 Assignment - Part 3

Explore and cluster the neighborhoods in Toronto.
Generate maps to visualize your neighborhoods and how they cluster together.

# *** Note: The map, if unable to be seen here in this notebook, can also be viewed separately as a labeled file in GitHub.

*** Also, the Clustering Analysis is located towards the end of this notebook. 

# *** Please keep scrolling down the page because the Toronto & Cluster maps take up a lot of space.

*** The Clustering Analysis can be located when scrolling is done. Thank you.

In [1]:
import pandas as pd
import requests

# Scrape the list of Postal Codes of Canada along with Borough and Neighbourhood.

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [3]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

Data Cleansing: One Postal Code can have multiple Neighbourhoods.

Data Cleansing: Determine Postal Codes along with Boroughs & Neighborhoods that are 'Not assigned'.

Data Cleansing: If a Borough has a 'Not assigned' Neighbourhood,then the Neighbourhood is the same as the Borough.

In [4]:
len(wiki_data), type(wiki_data)

(3, list)

In [5]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

# Data Cleansing: Only rows with an assigned borough are processed; if 'Not assigned', these are ignored (removed)from the list.

In [6]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

# Data Cleansing: Postal Codes can have multiple Neighbourhoods; the 'Not assigned' ones have been removed.

In [7]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [9]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

# There are 103 rows covering the data cleansed Postal Codes, Boroughs, and Neighbourhoods.

In [11]:
df.shape

(103, 3)

# The next segment involves determining the latitude and longitude for the list of Neighbourhoods by Postal Code and Borough.

In [12]:
pip install geocoder

     |████████████████████████████████| 98 kB 8.5 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder

In [14]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [15]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [17]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [18]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [19]:
combined_data.shape

(103, 5)

# Note: The sections above represent the work completed for Parts 1 (Data Cleansing) & 2 (Latitude, Longitude) of the Week 3 Assignment.

# Part 3 of Week 3 Assignment: The sections below represent the continuation of the assignment, Part 3, by focusing on clustering and mapping the information for analysis.

Preparing the data for Map1: Toronto to see the latitude and longitude information.

Note: If the Toronto map is not seen here in this Notebook, please refer also to a separate file in GitHub.

# *When viewing the map in the Notebook, it creates a huge blank space; PLEASE CONTINUE SCROLLING DOWN THE PAGE for the rest of the assignment.

In [20]:
from geopy.geocoders import Nominatim

In [21]:
pip install folium

     |████████████████████████████████| 94 kB 5.3 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [26]:
address = 'Toronto, ON'
geolocator = Nominatim(user_agent="Toronto")
location = geolocator.geocode(address)
latitude_toronto = location.latitude
longitude_toronto = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude_toronto, longitude_toronto))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [27]:
import folium
# Creating the map of Toronto
map_toronto = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)

map_toronto

# Explore Toronto Neighbourhoods:
Define Foursquare Credentials and Version

In [28]:
# @hidden_cell
CLIENT_ID = '1EVBZSDTXZBSLMZURSV2F1HHL2PDL1T454KB0CLTZJNVZWZ4' # your Foursquare ID
CLIENT_SECRET = 'TUE1P5CL1VI315KF0OVBXPT1PADLJB3GE41GTWTXOHAYGYQN' # your Foursquare Secret
VERSION = '20200605' # Foursquare API version
print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: 1EVBZSDTXZBSLMZURSV2F1HHL2PDL1T454KB0CLTZJNVZWZ4
CLIENT_SECRET:TUE1P5CL1VI315KF0OVBXPT1PADLJB3GE41GTWTXOHAYGYQN


# Establish criteria (names, latitude, longitude, radius) to identify nearby venues by neighbourhoods in Toronto.

In [29]:
# defining radius and limit of venues to get
radius=500
LIMIT=100

In [38]:
import requests

In [40]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )

        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']

        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            lat,
            lng,
            v['venue']['name'],
            v['venue']['location']['lat'],
            v['venue']['location']['lng'],
            v['venue']['categories'][0]['name']) for v in results])
    
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood',
                'Neighbourhood Latitude',
                'Neighbourhood Longitude',
                'Venue',
                'Venue Latitude',
                'Venue Longitude',
                'Venue Category']

    return(nearby_venues)

In [42]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

# Identified nearby venues by Neighbourhood.

In [43]:
venues_in_toronto.head(10)

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0                  Parkwoods               43.753259               -79.329656   
1                  Parkwoods               43.753259               -79.329656   
2                  Parkwoods               43.753259               -79.329656   
3           Victoria Village               43.725882               -79.315572   
4           Victoria Village               43.725882               -79.315572   
5           Victoria Village               43.725882               -79.315572   
6           Victoria Village               43.725882               -79.315572   
7  Regent Park, Harbourfront               43.654260               -79.360636   
8  Regent Park, Harbourfront               43.654260               -79.360636   
9  Regent Park, Harbourfront               43.654260               -79.360636   

                                        Venue  Venue Latitude  \
0                             Brookbanks Park       43.751976   
1                               Variety Store       43.751974   
2  Bella Vita Catering & Private Chef Service       43.756651   
3                      Victoria Village Arena       43.723481   
4                                   Portugril       43.725819   
5                                 Tim Hortons       43.725517   
6                                    The Frig       43.727051   
7                            Roselle Desserts       43.653447   
8                               Tandem Coffee       43.653559   
9                      Cooper Koo Family YMCA       43.653249   

   Venue Longitude         Venue Category  
0       -79.332140                   Park  
1       -79.333114      Food & Drink Shop  
2       -79.331524              BBQ Joint  
3       -79.315635           Hockey Arena  
4       -79.312785  Portuguese Restaurant  
5       -79.313103            Coffee Shop  
6       -79.317418      French Restaurant  
7       -79.362017                 Bakery  
8       -79.361809            Coffee Shop  
9       -79.358008    Distribution Center

In [44]:
venues_in_toronto.shape

(1317, 7)

In [45]:
venues_in_toronto.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    22   
...                                                                 ...   
Willowdale, Willowdale West                                           4   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       2   
York Mills, Silver Hills                                              1   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                              4   
Alderwood, Long Branch                                                 8   
Bathurst Manor, Wilson Heights, Downsview North                       21   
Bayview Village                                                        4   
Bedford Park, Lawrence Manor East                                     22   
...                                                                  ...   
Willowdale, Willowdale West                                            4   
Woburn                                                                 4   
Woodbine Heights                                                       7   
York Mills West                                                        2   
York Mills, Silver Hills                                               1   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   22              22   
...                                                ...             ...   
Willowdale, Willowdale West                          4               4   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      2               2   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             22   
...                                                          ...   
Willowdale, Willowdale West                                    4   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                2   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     4  
Alderwood, Long Branch                

# Analyze each neighborhood

In [46]:
# one hot encoding
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat.head()

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Vegetarian / Vegan Restaurant  \
0             0         0  ...                              0   
1             0         0  ...                              0   
2             0         0  ...                              0   
3             0         0  ...                              0   
4             0         0  ...                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio     Neighbourhood  
0         0            0              0            0         Parkwoods  
1         0            0              0            0         Parkwoods  
2         0            0              0            0         Parkwoods  
3         0            0              0            0  Victoria Village  
4         0            0              0            0  Victoria Village  

[5 rows x 239 columns]

In [47]:
toronto_venue_cat.shape

(1317, 239)

# Determine the Mean Frequency of the venue categories by Neighbourhood.

In [48]:
#Next, group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
92                      Willowdale, Willowdale West                0.0   
93                                           Woburn                0.0   
94                                 Woodbine Heights                0.0   
95                                  York Mills West                0.0   
96                         York Mills, Silver Hills                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
1       0.0                 0.0           0.0             0.0   
2       0.0                 0.0           0.0             0.0   
3       0.0                 0.0           0.0             0.0   
4       0.0                 0.0           0.0             0.0   
..      ...                 ...           ...             ...   
92      0.0                 0.0           0.0             0.0   
93      0.0                 0.0           0.0             0.0   
94      0.0                 0.0           0.0             0.0   
95      0.0                 0.0           0.0             0.0   
96      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.045455           0.0  ...   
..              ...               ...                  ...           ...  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   
94              0.0               0.0             0.000000           0.0  ...   
95              0.0               0.0             0.000000           0.0  ...   
96              0.0               0.0             0.000000           0.0  ...   

    Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                  0.0                            0.0               0.0   
1                  0.0                            0.0               0.0   
2                  0.0                            0.0               0.0   
3                  0.0                            0.0               0.0   
4                  0.0                            0.0               0.0   
..                 ...                            ...               ...   
92                 0.0                            0.0               0.0   
93                 0.0                            0.0               0.0   
94                 0.0                            0.0               0.0   
95                 0.0                            0.0               0.0   
96                 0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000                    0.0              0.0       0.0   
1      0.000000                    0.0              0.0       0.0   
2      0.000000                    0.0              0.0       0.0   
3      0.000000                    0.0              0.0       0.0   
4      0.000000                    0.0              0.0       0.0   
..          ...                    ...              ...       ...   
92     0.00

# Print each neighborhood along with the top 5 most common venues.

In [49]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4          Accessories Store  0.00


----Alderwood, Long Branch----
          venue  freq
0   Pizza Place  0.25
1           Gym  0.12
2   Coffee Shop  0.12
3           Pub  0.12
4  Dance Studio  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                       Bank  0.10
1                Coffee Shop  0.10
2                   Pharmacy  0.05
3                Gas Station  0.05
4  Middle Eastern Restaurant  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3                 Café  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0      Sandwich Place  0.09
1  Italian Restaurant  0.09
2         C



----North Park, Maple Leaf Park, Upwood Park----
                        venue  freq
0              Massage Studio  0.25
1  Construction & Landscaping  0.25
2                        Park  0.25
3                      Bakery  0.25
4           Accessories Store  0.00


----North Toronto West, Lawrence Park----
                  venue  freq
0        Clothing Store  0.16
1   Sporting Goods Shop  0.11
2           Coffee Shop  0.11
3                   Spa  0.05
4  Fast Food Restaurant  0.05


----Northwest, West Humber - Clairville----
                 venue  freq
0        Garden Center  0.25
1  Rental Car Location  0.25
2           Truck Stop  0.25
3            Drugstore  0.25
4    Accessories Store  0.00


----Northwood Park, York University----
                    venue  freq
0  Furniture / Home Store  0.14
1             Coffee Shop  0.14
2      Miscellaneous Shop  0.14
3                     Bar  0.14
4          Massage Studio  0.14


----Old Mill South, King's Mill Park, Sunnylea, Humbe

The venues are first sorted in descending order and then the above data is put into a dataframe.

In [50]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

The new dataframe is created to display the top 10 venues for each Neighbourhood.

In [51]:
#Now let's create the new dataframe and display the top 10 venues for each neighborhood.

import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village                       Café   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot          Skating Rink                Lounge   
1        Sandwich Place           Coffee Shop                   Pub   
2                  Bank             Pet Store        Ice Cream Shop   
3    Chinese Restaurant                  Bank   Japanese Restaurant   
4           Coffee Shop    Italian Restaurant         Grocery Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio          Dessert Shop           Escape Room   
1          Dance Studio              Pharmacy                   Gym   
2         Shopping Mall           Bridal Shop                 Diner   
3      Department Store           Escape Room     Electronics Store   
4      Toy / Game Store             Juice Bar            Restaurant   

         8th Most Common Venue        9th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
1                    Gastropub             Cuban Restaurant   
2               Sandwich Place                Deli / Bodega   
3  Eastern European Restaurant                    Drugstore   
4             Sushi Restaurant            Indian Restaurant   

  10th Most Common Venue  
0              Drugstore  
1                Dog Run  
2            Supermarket  
3             Donut Shop  
4                Butcher

# Cluster Neighborhoods
Run k-means to cluster the neighborhood into 5 clusters.

In [52]:
# import k-means from clustering stage 
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 2, 1, 2, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1,
       2, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 3,
       0, 2, 1, 1, 0, 1, 1, 2, 2], dtype=int32)

A new dataframe is created that includes the cluster along with top 10 venue for each Neighbourhood.

In [53]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

toronto_merged = combined_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster_Labels 1st Most Common Venue  \
0  43.753259 -79.329656             2.0                  Park   
1  43.725882 -79.315572             1.0     French Restaurant   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop              BBQ Joint           Yoga Studio   
1           Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                  Park                 Bakery        Breakfast Spot   
3     Accessories Store  Vietnamese Restaurant              Boutique   
4      Sushi Restaurant                    Gym             Nightclub   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Department Store           Escape Room     Electronics Store   
1          Discount Store      Department Store          Dessert Shop   
2                    Café            Restaurant                   Pub   
3  Furniture / Home Store    Athletics & Sports           Event Space   
4                Beer Bar         Smoothie Shop        Sandwich Place   

         8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant              Drugstore             Donut Shop  
1           Dim Sum Restaurant                  Diner            Yoga Studio  
2               Chocolate Shop  Performing Arts Venue           Dessert Shop  
3                  Coffee Shop           Dessert Shop            Escape Room  
4                Burrito Place                   Café  Portuguese Restaurant

In [54]:
#We find that there is no data available for some neighbourhood droping that row

toronto_merged=toronto_merged.dropna()

In [55]:
toronto_merged['Cluster_Labels'] = toronto_merged.Cluster_Labels.astype(int)

In [ ]:
NOTE: The section below further involves a 2nd Map: "Map Cluster".

# If the Map Cluster is not seen in the Notebook, please refer to a separately labelled map file in GitHub.
***Please KEEP SCROLLING DOWN THE PAGE for the Cluster Analysis as the MAP leave a huge space on the page.
****PLEASE KEEP SCROLLING for more of the assignment work. Thank you.

In [56]:
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude_toronto, longitude_toronto], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Cluster 1 with "Cluster_Label 0"

In [57]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue   2nd Most Common Venue  \
5    Etobicoke               0           Pizza Place             Event Space   
50  North York               0           Pizza Place  Furniture / Home Store   
63        York               0     Convenience Store                Bus Line   
70   Etobicoke               0           Pizza Place             Coffee Shop   
72  North York               0         Grocery Store             Coffee Shop   
93   Etobicoke               0           Pizza Place          Sandwich Place   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
5   Ethiopian Restaurant           Escape Room          Electronics Store   
50          Intersection         Deli / Bodega                Escape Room   
63           Pizza Place           Yoga Studio           Department Store   
70        Discount Store    Chinese Restaurant  Middle Eastern Restaurant   
72              Pharmacy           Pizza Place               Airport Gate   
93           Coffee Shop                   Pub               Dance Studio   

          6th Most Common Venue        7th Most Common Venue  \
5   Eastern European Restaurant                    Drugstore   
50            Electronics Store  Eastern European Restaurant   
63                  Escape Room            Electronics Store   
70               Sandwich Place                 Intersection   
72             Department Store         Ethiopian Restaurant   
93                     Pharmacy                          Gym   

          8th Most Common Venue        9th Most Common Venue  \
5                    Donut Shop                      Dog Run   
50                    Drugstore                   Donut Shop   
63  Eastern European Restaurant                    Drugstore   
70             Department Store  Eastern European Restaurant   
72                  Escape Room            Electronics Store   
93                    Gastropub             Cuban Restaurant   

         10th Most Common Venue  
5           Distribution Center  
50                      Dog Run  
63                   Donut Shop  
70                    Drugstore  
72  Eastern European Restaurant  
93                      Dog Run

# Examine Cluster 2 with "Cluster_Label 1"

In [58]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels       1st Most Common Venue  \
1          North York               1           French Restaurant   
2    Downtown Toronto               1                 Coffee Shop   
3          North York               1              Clothing Store   
4    Downtown Toronto               1                 Coffee Shop   
6         Scarborough               1        Fast Food Restaurant   
..                ...             ...                         ...   
97   Downtown Toronto               1                  Restaurant   
99   Downtown Toronto               1                 Coffee Shop   
100      East Toronto               1          Light Rail Station   
101         Etobicoke               1  Construction & Landscaping   
102         Etobicoke               1               Grocery Store   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop  Portuguese Restaurant          Hockey Arena   
2                      Park                 Bakery        Breakfast Spot   
3         Accessories Store  Vietnamese Restaurant              Boutique   
4          Sushi Restaurant                    Gym             Nightclub   
6                Print Shop            Yoga Studio         Deli / Bodega   
..                      ...                    ...                   ...   
97                     Café            Coffee Shop    Seafood Restaurant   
99                     Café    Japanese Restaurant                   Pub   
100           Auto Workshop             Comic Shop           Pizza Place   
101          Baseball Field            Yoga Studio          Dessert Shop   
102  Thrift / Vintage Store      Convenience Store        Discount Store   

      5th Most Common Venue        6th Most Common Venue  \
1            Discount Store             Department Store   
2                      Café                   Restaurant   
3    Furniture / Home Store           Athletics & Sports   
4                  Beer Bar                Smoothie Shop   
6         Electronics Store  Eastern European Restaurant   
..                      ...                          ...   
97              Art Gallery                    Speakeasy   
99         Ramen Restaurant               Ice Cream Shop   
100              Restaurant                      Butcher   
101    Ethiopian Restaurant                  Escape Room   
102            Burger Joint               Sandwich Place   

    7th Most Common Venue        8th Most Common Venue  9th Most Common Venue  \
1            Dessert Shop           Dim Sum Restaurant                  Diner   
2                     Pub               Chocolate Shop  Performing Arts Venue   
3             Event Space                  Coffee Shop           Dessert Shop   
4          Sandwich Place                Burrito Place                   Café   
6               Drugstore                   Donut Shop                Dog Run   
..                    ...                          ...                    ...   
97                    Pub                   Steakhouse                  Hotel   
99       Sushi Restaurant            Indian Restaurant               Beer Bar   
100         Burrito Place                   Skate Park                Brewery   
101     Electronics Store  Eastern European Restaurant              Drugstore   
102  Fast Food Restaurant              Supplement Shop                 Bakery   

    10th Most Common Venue  
1              Yoga Studio  
2             Dessert Shop  
3              Escape Room  
4    Portuguese Restaurant  
6      Distribution Center  
..                     ...  
97                  Bakery  
99    Ethiopian Restaurant  
100             Smoke Shop  
101             Donut Shop  
102          Tanning Salon  

[82 rows x 12 columns]

# Examine Cluster 3 with "Cluster_Label 2"

In [59]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue  \
0         North York               2                  Park   
21              York               2         Women's Store   
35         East York               2                  Park   
45        North York               2                  Park   
49        North York               2                  Park   
61   Central Toronto               2      Business Service   
64              York               2                  Park   
66        North York               2                  Park   
77         Etobicoke               2                  Park   
85       Scarborough               2          Intersection   
91  Downtown Toronto               2                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop             BBQ Joint           Yoga Studio   
21                        Park                  Pool         Deli / Bodega   
35           Convenience Store          Intersection           Yoga Studio   
45                 Yoga Studio         Deli / Bodega           Escape Room   
49  Construction & Landscaping        Massage Studio                Bakery   
61                        Park           Swim School              Bus Line   
64                 Yoga Studio         Deli / Bodega           Escape Room   
66           Convenience Store           Yoga Studio      Department Store   
77           Mobile Phone Shop        Sandwich Place           Yoga Studio   
85                  Playground                  Park   Arts & Crafts Store   
91                  Playground                 Trail           Yoga Studio   

   5th Most Common Venue        6th Most Common Venue  \
0       Department Store                  Escape Room   
21           Escape Room            Electronics Store   
35      Department Store                  Escape Room   
45     Electronics Store  Eastern European Restaurant   
49           Yoga Studio                 Dessert Shop   
61           Yoga Studio                  Escape Room   
64     Electronics Store  Eastern European Restaurant   
66           Escape Room            Electronics Store   
77      Department Store            Electronics Store   
85           Yoga Studio               Discount Store   
91        Discount Store             Department Store   

          7th Most Common Venue        8th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21  Eastern European Restaurant                    Drugstore   
35            Electronics Store  Eastern European Restaurant   
45                    Drugstore                   Donut Shop   
49                  Escape Room            Electronics Store   
61            Electronics Store  Eastern European Restaurant   
64                    Drugstore                   Donut Shop   
66  Eastern European Restaurant                    Drugstore   
77  Eastern European Restaurant                    Drugstore   
85                 Dessert Shop           Dim Sum Restaurant   
91                 Dessert Shop           Dim Sum Restaurant   

          9th Most Common Venue 10th Most Common Venue  
0                     Drugstore             Donut Shop  
21                   Donut Shop                Dog Run  
35                    Drugstore             Donut Shop  
45                      Dog Run    Distribution Center  
49  Eastern European Restaurant              Drugstore  
61                    Drugstore             Donut Shop  
64                      Dog Run    Distribution Center  
66                   Donut Shop                Dog Run  
77                   Donut Shop                Dog Run  
85                        Diner                Dog Run  
91                        Diner                Dog Run

# Examine Cluster 4 with "Cluster_Label 3"

In [60]:

toronto_merged.loc[toronto_merged['Cluster_Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke               3   Filipino Restaurant           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Fabric Shop  Ethiopian Restaurant           Escape Room   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
11     Electronics Store  Eastern European Restaurant             Drugstore   

   9th Most Common Venue 10th Most Common Venue  
11            Donut Shop                Dog Run

# Examine Cluster 5 with "Cluster_Label 4"

In [61]:
toronto_merged.loc[toronto_merged['Cluster_Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster_Labels 1st Most Common Venue 2nd Most Common Venue  \
98  Etobicoke               4                 River           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98         Deli / Bodega           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
98  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
98               Dog Run    Distribution Center

# As seen from the above analysis, Cluster 2 with "Cluster_Label 1", with its significant number of bouroughs (more than 80), offers the most options for diverse neighborhood venues.

# Many residents, along with visitors, can enjoy such activities as dining, running errands, shopping, working out, playing/watching sports, and relaxing in the park.